include/seqan3/range/view/kmer_hash.hpp

```diff
@@ -12,47 +12,645 @@
+         */
+        void hash_backward(difference_type const skip)
+        {
+            std::ranges::advance(text_left, -skip);
+            hash_full();
+        }
+
+        //!\brief Calculates a hash value by explicitly looking at each position.
+        void hash_full()
+        {
+            assert(pview != nullptr);
+
+            text_right = text_left;
+            hash_value = 0;
+
+            for (size_t i{0}; i < pview->shape_.size() - 1u; ++i)
+            {
+                hash_value += pview->shape_[i] * to_rank(*text_right);
+                hash_value *= pview->shape_[i] ? sigma : 1;
+                std::ranges::advance(text_right, 1);
+            }
+        }
+
+        //!\brief Calculates the next hash value via rolling hash.
+        void hash_roll_forward()
+        {
+            hash_value -= to_rank(*(text_left)) * roll_factor;
+            hash_value += to_rank(*(text_right));
+            hash_value *= sigma;
+
+            std::ranges::advance(text_left,  1);
+            std::ranges::advance(text_right, 1);
+        }
+
+        /*!\brief Calculates the previous hash value via rolling hash.
+         * \attention This function is only avaible if `it_t` models std::BidirectionalIterator.
+         */
+        void hash_roll_backward()
+            //!\cond
+            requires std::BidirectionalIterator<it_t>
+            //!\endcond
+        {
+            std::ranges::advance(text_left,  -1);
+            std::ranges::advance(text_right, -1);
+
+            hash_value /= sigma;
+            hash_value -= to_rank(*(text_right));
+            hash_value += to_rank(*(text_left)) * roll_factor;
+        }
+    };
+
+public:
+    /*!\name Associated types
+     * \{
+     */
+    //!\brief The iterator type of this view.
+    using iterator          = shape_iterator<urng_t>;
+    //!\brief The const_iterator type of this view.
+    using const_iterator    = shape_iterator<urng_t const>;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    kmer_hash_view()                                       = default; //!< Defaulted.
+    kmer_hash_view(kmer_hash_view const & rhs)             = default; //!< Defaulted.
+    kmer_hash_view(kmer_hash_view && rhs)                  = default; //!< Defaulted.
+    kmer_hash_view & operator=(kmer_hash_view const & rhs) = default; //!< Defaulted.
+    kmer_hash_view & operator=(kmer_hash_view && rhs)      = default; //!< Defaulted.
+    ~kmer_hash_view()                                      = default; //!< Defaulted.
+
+    //!\brief Construct from a view and a given shape.
+    kmer_hash_view(urng_t urange_, shape const & s_) : urange{std::move(urange_)}, shape_{s_}
+    {}
+
+    //!\brief Construct from a non-view that can be view-wrapped and a given shape.
+    template <std::ranges::ViewableRange orng_t>
+    kmer_hash_view(orng_t && urange_, shape const & s_) :
```

Comment:
Add some extra checks here that you can actually construct from the view. See other views for reference.